File: UIScrollView-Actions/UIScrollView+Actions.m
//
//  UIScrollView+Actions.m
//  CALX
//
//  Created by Daniel Clelland on 14/12/13.
//  Copyright (c) 2013 Daniel Clelland. All rights reserved.
//

#import "UIScrollView+Actions.h"

@implementation UIScrollView (Actions)

- (CGPoint)topContentOffset
{
    return CGPointMake(0.0f, -self.contentInset.top);
}

- (CGPoint)bottomContentOffset
{
    return CGPointMake(0.0f, self.contentSize.height + self.contentInset.bottom - self.bounds.size.height);
}

- (CGPoint)leftContentOffset
{
    return CGPointMake(-self.contentInset.left, 0.0f);
}

- (CGPoint)rightContentOffset
{
    return CGPointMake(self.contentSize.width + self.contentInset.right - self.bounds.size.width, 0.0f);
}

- (BOOL)isScrolledToTop
{
    return self.contentOffset.y <= [self topContentOffset].y;
}

- (BOOL)isScrolledToBottom
{
    return self.contentOffset.y >= [self bottomContentOffset].y;
}

- (BOOL)isScrolledToLeft
{
    return self.contentOffset.x <= [self leftContentOffset].x;
}

- (BOOL)isScrolledToRight
{
    return self.contentOffset.x >= [self rightContentOffset].x;
}

- (void)scrollToTopAnimated:(BOOL)animated
{
    [self setContentOffset:[self topContentOffset] animated:animated];
}

- (void)scrollToBottomAnimated:(BOOL)animated
{
    [self setContentOffset:[self bottomContentOffset] animated:animated];
}

- (void)scrollToLeftAnimated:(BOOL)animated
{
    [self setContentOffset:[self leftContentOffset] animated:animated];
}

- (void)scrollToRightAnimated:(BOOL)animated
{
    [self setContentOffset:[self rightContentOffset] animated:animated];
}

- (NSUInteger)verticalPageIndex
{
    return (self.contentOffset.y + (self.frame.size.height * 0.5f)) / self.frame.size.height;
}

- (NSUInteger)horizontalPageIndex
{
    return (self.contentOffset.x + (self.frame.size.width * 0.5f)) / self.frame.size.width;
}

- (void)scrollToVerticalPageIndex:(NSUInteger)pageIndex animated:(BOOL)animated
{
    [self setContentOffset:CGPointMake(0.0f, self.frame.size.height * pageIndex) animated:animated];
}

- (void)scrollToHorizontalPageIndex:(NSUInteger)pageIndex animated:(BOOL)animated
{
    [self setContentOffset:CGPointMake(self.frame.size.width * pageIndex, 0.0f) animated:animated];
}

@end

@implementation UIViewController (UIScrollView)

- (UIScrollView *)scrollView
{
    if ([self isKindOfClass:[UITableViewController class]]) {
        UITableViewController *tableViewController = (UITableViewController *)self;
        return tableViewController.tableView;
    }
    
    if ([self isKindOfClass:[UICollectionViewController class]]) {
        UICollectionViewController *collectionViewController = (UICollectionViewController *)self;
        return collectionViewController.collectionView;
    }
    
    return nil;
}

@end

File: UIScrollView-Actions/UIScrollView+Actions.swift
//
//  UIScrollView+Actions.swift
//  Pods
//
//  Created by Daniel Clelland on 11/03/17.
//
//

import UIKit

// MARK: Scrolling

public extension UIScrollView {
    
    /// A scroll view edge.
    public enum Edge {
        
        /// The scroll view's top edge.
        case top
        
        /// The scroll view's right edge.
        case right
        
        /// The scroll view's bottom edge.
        case bottom
        
        /// The scroll view's left edge.
        case left
    }
    
    /// Scrolls the scroll view to the specified edge.
    ///
    /// Only one coordinate is modified; if you scroll to the `.left` edge, the `contentOffset`'s `y` value won't be modified.
    ///
    /// If you are using content insets, they are respected, and the scroll view won't scroll too far.
    public func scroll(to edge: Edge, animated: Bool = false) {
        setContentOffset(contentOffset(for: edge), animated: animated)
    }
    
    /// Checks if the scroll view is scrolled to (or past) a given edge.
    ///
    /// Content insets are respected.
    public func isScrolled(to edge: Edge) -> Bool {
        switch edge {
        case .top:
            return contentOffset.y <= contentOffset(for: edge).y
        case .right:
            return contentOffset.x >= contentOffset(for: edge).x
        case .bottom:
            return contentOffset.y >= contentOffset(for: edge).y
        case .left:
            return contentOffset.x <= contentOffset(for: edge).x
        }
    }
    
    /// Calculates the correct content offset for a given edge.
    private func contentOffset(for edge: Edge) -> CGPoint {
        switch edge {
        case .top:
            return CGPoint(
                x: contentOffset.x,
                y: -contentInset.top
            )
        case .right:
            return CGPoint(
                x: contentSize.width + contentInset.right - bounds.width,
                y: contentOffset.y
            )
        case .bottom:
            return CGPoint(
                x: contentOffset.x,
                y: contentSize.height + contentInset.bottom - bounds.height
            )
        case .left:
            return CGPoint(
                x: -contentInset.left,
                y: contentOffset.y
            )
        }
    }
    
}

// MARK: Pagination

public extension UIScrollView {
    
    /// A scroll view page index on either the horizontal or vertical axis.
    public enum Page {
        
        /// A page index on the horizontal axis.
        case horizontal(index: Int)
        
        /// A page index on the vertical axis.
        case vertical(index: Int)
    }
    
    /// Scrolls the scroll view to the specified pagination location.
    ///
    /// Pages are considered to be the height or width of the scroll view's frame.
    ///
    /// Only one coordinate is modified; if you scroll to a `.horizontal` page index, the `contentOffset`'s `y` value won't be modified.
    ///
    /// If you are using content insets, they are respected, and the scroll view won't scroll too far.
    public func scroll(to page: Page, animated: Bool = false) {
        setContentOffset(contentOffset(for: page), animated: animated)
    }
    
    /// Calculates the correct content offset for a given page index.
    private func contentOffset(for page: Page) -> CGPoint {
        switch page {
        case .horizontal(let page):
            return CGPoint(
                x: CGFloat(page) * frame.width - contentInset.left,
                y: contentOffset.y
            )
        case .vertical(let page):
            return CGPoint(
                x: contentOffset.x,
                y: CGFloat(page) * frame.height - contentInset.top
            )
        }
    }
    
}
